File: src/main/java/org/apache/prepbuddy/cleansers/imputation/UnivariateLinearRegressionSubstitution.java
package org.apache.prepbuddy.cleansers.imputation;

import org.apache.commons.lang.math.NumberUtils;
import org.apache.prepbuddy.rdds.TransformableRDD;
import org.apache.prepbuddy.transformers.RowPurger;
import org.apache.prepbuddy.utils.RowRecord;
import org.apache.spark.api.java.JavaDoubleRDD;

import java.text.DecimalFormat;

/**
 * An imputation strategy that is based on Linear Regression which is an approach
 * for modeling the relationship between a scalar dependent variable y and an explanatory
 * variable x.
 * This strategy imputes the value of y by : slope * x + intercept
 */
public class UnivariateLinearRegressionSubstitution implements ImputationStrategy {

    private static final String BLANK_STRING = "";
    private int independentColumnIndex;
    private double slope;
    private double intercept;

    public UnivariateLinearRegressionSubstitution(int independentColumnIndex) {
        this.independentColumnIndex = independentColumnIndex;
    }

    @Override
    public void prepareSubstitute(TransformableRDD inputDataset, final int missingDataColumnIndex) {
        TransformableRDD rddForRegression = inputDataset.removeRows(new RowPurger.Predicate() {
            @Override
            public Boolean evaluate(RowRecord record) {
                String xColumnValue = record.valueAt(independentColumnIndex);
                String yColumnValue = record.valueAt(missingDataColumnIndex);
                return !NumberUtils.isNumber(xColumnValue) || xColumnValue.trim().isEmpty() || yColumnValue.trim().isEmpty();
            }
        });
        long count = rddForRegression.count();

        JavaDoubleRDD xyRDD = rddForRegression.multiplyColumns(missingDataColumnIndex, independentColumnIndex);
        JavaDoubleRDD xSquaredRDD = rddForRegression.multiplyColumns(independentColumnIndex, independentColumnIndex);
        JavaDoubleRDD yRDD = rddForRegression.toDoubleRDD(missingDataColumnIndex);
        JavaDoubleRDD xRDD = rddForRegression.toDoubleRDD(independentColumnIndex);

        Double squareRddSum = xSquaredRDD.sum();
        Double sumOfXY = xyRDD.sum();
        Double sumOfY = yRDD.sum();
        Double sumOfX = xRDD.sum();

        setSlope(sumOfX, sumOfY, sumOfXY, squareRddSum, count);
        setIntercept(sumOfX, sumOfY, count);
    }

    private void setIntercept(Double sumOfXs, Double sumOfYs, long count) {
        intercept = (sumOfYs - (slope * sumOfXs)) / count;
    }

    private void setSlope(Double sumOfXs, Double sumOfYs, Double sumOfXYs, Double sumOfXSquared, long count) {
        slope = ((count * sumOfXYs) - (sumOfXs * sumOfYs)) / ((count * sumOfXSquared) - sumOfXs * sumOfXs);
    }

    @Override
    public String handleMissingData(RowRecord record) {
        String dependentValue = record.valueAt(independentColumnIndex);
        try {
            Double value = Double.parseDouble(dependentValue);
            Double imputedValue = intercept + slope * value;
            imputedValue = Double.parseDouble(new DecimalFormat("##.##").format(imputedValue));
            return imputedValue.toString();
        } catch (NumberFormatException e) {
            return BLANK_STRING;
        }
    }
}

File: src/main/scala/org/apache/datacommons/prepbuddy/imputations/UnivariateLinearRegressionSubstitution.scala
package org.apache.datacommons.prepbuddy.imputations

import org.apache.datacommons.prepbuddy.rdds.TransformableRDD
import org.apache.datacommons.prepbuddy.utils.RowRecord
import org.apache.spark.rdd.RDD

/**
  * An imputation strategy that is based on Linear Regression which is an approach
  * for modeling the relationship between a scalar dependent variable y and an explanatory
  * variable x.
  * This strategy imputes the value of y by : slope * x + intercept
  * This implementation is only for imputing numeric columns.
  */
class UnivariateLinearRegressionSubstitution(independentColumn: Int) extends ImputationStrategy {
    private var slope: Double = 0
    private var intercept: Double = 0

    override def prepareSubstitute(tRdd: TransformableRDD, missingDataColumn: Int): Unit = {
        val rdd: TransformableRDD = tRdd.removeRows((row) => {
            row.select(missingDataColumn).trim.isEmpty || row.select(independentColumn).trim.isEmpty
        })
        val xyRDD: RDD[Double] = rdd.multiplyColumns(missingDataColumn, independentColumn)
        val xSquareRDD: RDD[Double] = rdd.multiplyColumns(independentColumn, independentColumn)
        val yRDD: RDD[Double] = rdd.toDoubleRDD(missingDataColumn)
        val xRDD: RDD[Double] = rdd.toDoubleRDD(independentColumn)

        val sumOfXY: Double = xyRDD.sum()
        val sumOfSquareRDD: Double = xSquareRDD.sum()
        val sumOfX: Double = xRDD.sum()
        val sumOfY: Double = yRDD.sum()

        val count: Long = xSquareRDD.count()
        setSlope(sumOfX, sumOfY, sumOfXY, sumOfSquareRDD, count)
        setIntercept(sumOfX, sumOfY, count)
    }

    def setSlope(sumOfXs: Double, sumOfYs: Double, sumOfXYs: Double, sumOfSquared: Double, count: Long): Unit = {
        val nominator: Double = (count * sumOfXYs) - (sumOfXs * sumOfYs)
        val denominator: Double = (count * sumOfSquared) - sumOfXs * sumOfXs
        slope = nominator / denominator
    }

    def setIntercept(sumOfXs: Double, sumOfYs: Double, count: Long): Unit = {
        intercept = (sumOfYs - (slope * sumOfXs)) / count
    }

    override def handleMissingData(record: RowRecord): String = {
        val independentValue: String = record.select(independentColumn)
        try {
            val value: Double = independentValue.toDouble
            val imputedValue: Double = intercept + slope * value
            "%1.2f".format(imputedValue)
        }
        catch {
            case exp: NumberFormatException => ""
        }
    }
}
